### ISSUE TITLE
Token collision in JwkTokenStore with multiple issuers

### ISSUE THREAD

**Issue by daniellavoie** [spring-projects/spring-security-oauth#1178]
Acceptance Tests on Spring Cloud Services highlighted an edge case that is outside of the JWT spec. `JwkTokenStore` support tokens from multiple providers.

However the `org.springframework.security.oauth2.provider.token.store.jwk.JwkDefinitionSource.jwkDefinitions` map the tokens from all the providers by `kid` from each token. 

Our test sets up two issuers that generate token with duplicate `kid`. This is not a real-world case but I don't think that the JWT spec covers that scenario. 

At the very least, logging the key collision when `JwkDefinitionSource` would be helpful for investigation. The symptom we encountered reflected itself as an `InvalidSignatureException` during the token's decoding.

I am no expert but would suggest supporting duplicate `kid` as Spring Security has no control on the issuers on which it depends. We implemented a workaround of our own by forking `JwkDefinitionSource` so tokens are grouped by `kid`rather than overwritten. Finally, the decode operation now tries all available token before failing.

At your disposal for furter details.

**Comment by jgrandja** [comment]
@daniellavoie We cannot go with the solution of _grouping_ `kid` 's in the `Map` and trying each `SignatureVerifier` associated to each `kid`. The guessing game is very dangerous so there needs to be a more intelligent key selection algorithm in place instead of just based on `kid`.

> At the very least, logging the key collision when JwkDefinitionSource reloads its cache would be helpful for investigation.

Agreed, this should be logged and likely a fail-fast approach implemented as well that clears the cache. Need to think about this one a bit though.


Can you please provide more details on your test case. Specifically, the JOSE headers for each of JWT's from the 2 providers.

The fact that 2 different provider's produce identical kid's is definitely and edge case and seem to me a vulnerability on the Provider side. Would you happen to know how PCF has implemented the `kid` value generation? I feel that it should at least include the `iss` attribute in combination with other provider attributes when producing a hash for the `kid` value in order to acheive uniqueness.

**Comment by daniellavoie** [comment]
I'll gather info for our sample tokens.

Meanwhile, I am concerned about the fact that the `kid` is an unspecified optional parameter according to [RFC 7515](https://tools.ietf.org/html/rfc7515#section-4.1.4). So whoever the provider is, we cannot make assumptions on it. I'm almost certain that PCF UAA generates a proper non-duplicate `kid` but we have no control on the providers used by our clients.

**Comment by daniellavoie** [comment]
After some digging, I came to realize that our Acceptance tests are running with a generic prod like PCF UAA. Here is an example of the key you would get from it : 

```
{
  "keys": [
    {
      "kty": "RSA",
      "e": "AQAB",
      "use": "sig",
      "kid": "key-1",
      "alg": "RS256",
      "value": "-----BEGIN PUBLIC KEY-----\nblablabla\n-----END PUBLIC KEY-----",
      "n": "blablabla"
    }
  ]
}
```

In our scenario, we get another key from another issuer (from another PCF env) with the same `kid`. So according to your explanation, PCF UAA would be vulnerable.

@habuma came out with an idea that could help us retrieving the proper key from the proper issuer. Tokens are encoded with an `iss` claim. That claim should match the issuers that were used when building the `JwkDefinitionSource.jwkDefinitions`. Another dimension could be added to this map to group `kid` by `issuer`. We would still have to loop through the keys and decode the token prior to retrieving the key from the cache, but at least a single `SignatureVerifier` call would be required.

Finally, the fail fast with logger error would be usefull for multi issuer scenario involving missing `kid` and `iss`.

**Comment by jgrandja** [comment]
@daniellavoie 
> Meanwhile, I am concerned about the fact that the `kid` is an unspecified optional parameter according to RFC 7515.

Yes, it is optional, but we know UAA generates this JOSE header because the `JwkTokenStore` was specifically developed for UAA and the key rotation feature.

> So whoever the provider is, we cannot make assumptions on it

We currently don't support providers that don't provide a `kid` in their Jwk Set or if it's missing in the JOSE header

**Comment by jgrandja** [comment]
@daniellavoie Let me think about your and @habuma proposed solution...

**Comment by jgrandja** [comment]
> In our scenario, we get another key from another issuer (from another PCF env) with the same kid. So according to your explanation, PCF UAA would be vulnerable.

I would raise this issue to the UAA team, as I feel the `kid` value should be unique across issuers - at least we can control it with our provider UAA. This can be done by using the issuer as one of the values when applying a MAC/Hash when generating the `kid` value.

**Comment by jgrandja** [comment]
@daniellavoie @habuma Regarding your proposal:

> That claim should match the issuers that were used when building the KwkDefinitionSource.jwkDefinitions

The problem is that the `JwkTokenStore` accepts JWK Set Url(s) which is different than the Issuer value so this would not work.

But yes ideally, we would like to group `kid` and `iss` as the key to the `JwkDefinitionHolder`.

**Comment by daniellavoie** [comment]
The JWK would be different but at least the domain would match. That could be a tradeoff.

**Comment by jgrandja** [comment]
The domain doesn't necessarily match. 
* JWK Set Url could be a sub-domain
* Jwk Set Url could be an unsecured endpoint, `http://`, whereas `iss` is required to be `https://`
* Jwk Set Url could technically be at any endpoint location and doesn't even need to look like the `iss` domain
